core/src/main/scala/cats/MonadError.scala

```diff
@@ -13,6 +13,23 @@ trait MonadError[F[_], E] extends ApplicativeError[F, E] with Monad[F] {
   def ensure[A](fa: F[A])(error: => E)(predicate: A => Boolean): F[A] =
     flatMap(fa)(a => if (predicate(a)) pure(a) else raiseError(error))
 
+  /**
+   * Sequences the specified finalizer ensuring evaluation regardless of
+   * whether or not the target `F[A]` raises an error.
+   *
+   * If `raiseError` is analogous to `throw` and `handleErrorWith` is analogous to
+   * `catch`, then `guarantee` is analogous to `finally`.  JVM exception semantics
+   * are mirrored with respect to error raised within the finalizer (i.e. errors
+   * raised within the finalizer will shade any errors raised by the primary action).
+   *
+   * @see [[raiseError]]
+   * @see [[handleErrorWith]]
+   */
+  def guarantee[A](fa: F[A], finalizer: F[Unit]): F[A] = {
```

Comment:
A law would be useless, since it would over-constrain implementations.  The only useful overrides of `guarantee` would be for implementations which have it as a structural primitive, but in that case `guarantee` would *not* be equivalent to `flatMap` composed with `attempt`.  So it would be a very round-about way of making the function `final` and running a ton of redundant tests.

The right thing to do is add a test, but unlike `MonadTest`, `MonadErrorTest` doesn't currently exist.  I'm all in favor of adding it, but this falls into the category of things that are somewhat harder to work on.  Incoming sometime in the next few days, hopefully.